### PR TITLE
Add gRPC for management of shadow shards

### PR DESCRIPTION
**Pull request by lferran**
### Description

This PR adds: 
 1. the `ShadowShardsManager` singleton that is responsible for managing the creation and deletion of shadow shards onto the node's disk.
  It is is also able to save brain resources operations and delete operations on disk.
  It will read all operations of a shadow shard in the same order they were inserted (by increasing txid).

 2. Added the grpc services on the sidecar and ingest to create and delete shadow shards.

NOTE: For the moment this grpc service isn't used yet. More PRs will come that use it

### How was this PR tested?
Unit and integration tests

### CODE REVIEW

**Comment by lferran** [comment]
[ch3780]

File: nucliadb_protos/nodewriter.proto [hunk 1]
```diff
@@ -77,8 +77,12 @@ message Counter {
...
 }
 
+message ShadowShardResponse {
+    bool success = 1;
+}
 
 service NodeSidecar {
-
     rpc GetCount(noderesources.ShardId) returns (Counter) {}
+    rpc ShadowShardCreate(noderesources.ShardId) returns (ShadowShardResponse) {}
```
Comment (lferran) [inline nucliadb_protos/nodewriter.proto:86]: Would it be better if the parameter would be an `EmptyQuery` and we would let the sidecar itself create the shadow shard with a random id?

File: nucliadb/nucliadb/ingest/service/writer.py [hunk 2]
```diff
@@ -741,6 +745,118 @@ async def generate_buffer(
         result = FileUploaded()
         return result
 
+    async def CreateShadowShard(  # type: ignore
```
Comment (bloodbare) [inline nucliadb/nucliadb/ingest/service/writer.py:748]: Should be created on the Node class, and not implemented on LocalNode

File: nucliadb/nucliadb/ingest/service/writer.py [hunk 3]
```diff
@@ -741,6 +745,118 @@ async def generate_buffer(
...
+                event_id = capture_exception(exc)
+            logger.error(
+                f"Error deleting shadow shard. Check sentry for more details. Event id: {event_id}"
+            )
+        else:
+            response.success = True
+        return response
+
 
 def update_shards_with_updated_replica(
```
Comment (bloodbare) [inline nucliadb/nucliadb/ingest/service/writer.py:861]: Move this code to Node

File: nucliadb_node/nucliadb_node/shadow_shards.py [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+            while line:
+                yield self.decode_operation(line)
+                line = await f.readline()
+
+
+def get_shadow_shards_manager() -> ShadowShardsManager:
+    global SHADOW_SHARDS_MANAGER
+
+    if SHADOW_SHARDS_MANAGER is None:
+        data_path = os.environ["DATA_PATH"]
```
Comment (bloodbare) [inline nucliadb_node/nucliadb_node/shadow_shards.py:249]: Better a get for error handling

File: nucliadb_node/nucliadb_node/shadow_shards.py [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+        SHADOW_SHARDS_MANAGER = ShadowShardsManager(
+            folder=f"{data_path}/shadow_shards/"
```
Comment (bloodbare) [inline nucliadb_node/nucliadb_node/shadow_shards.py:251]: As a constant

File: nucliadb_node/nucliadb_node/shadow_shards.py [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+            raise ShadowShardsNotLoaded()
+        if not self.exists(shard_id):
+            raise ShadowShardNotFound()
+
+        encoded = self.encode_operation((OperationCode.SET, brain))
+        shard_path = self.shard_path(shard_id)
+        async with aiofiles.open(shard_path, mode="ab") as f:
+            await f.write(encoded)
+
+        await self.increment_ops(shard_id)
```
Comment (bloodbare) [inline nucliadb_node/nucliadb_node/shadow_shards.py:216]: The incremental counting and the requirement to store it on each message is not useful, as the counting can be done reading the shard_path. Your implementation requires to write twice to dick on each message

File: nucliadb_node/nucliadb_node/shadow_shards.py [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+
+    def encode_operation(self, node_operation: NodeOperation) -> bytes:
+        opcode, payload = node_operation
+        if opcode == OperationCode.SET:
+            payload = payload.SerializeToString()  # type: ignore
+        elif opcode == OperationCode.DELETE:
+            payload = payload.encode()  # type: ignore
+        else:
+            ValueError(f"Unknown opcode: {opcode}")
+        encoded = base64.b64encode(payload).decode()  # type: ignore
```
Comment (bloodbare) [inline nucliadb_node/nucliadb_node/shadow_shards.py:193]: Why you encode in base64 ? We could just store the PBs on disc

File: nucliadb_node/nucliadb_node/shadow_shards.py [hunk 8]
```diff
@@ -0,0 +1,253 @@
...
+        await self.increment_ops(shard_id)
+
+    async def iter_operations(self, shard_id: str) -> AsyncIterator[NodeOperation]:
+        if not self.loaded:
+            raise ShadowShardsNotLoaded()
+        if not self.exists(shard_id):
+            raise ShadowShardNotFound()
+
+        shard_path = self.shard_path(shard_id)
+        async with aiofiles.open(shard_path, "r") as f:
```
Comment (bloodbare) [inline nucliadb_node/nucliadb_node/shadow_shards.py:238]: Instead of reading files we could iterate with glob

**Review by bloodbare — CHANGES_REQUESTED** [review]
Requires a refactor on storing the metadata and the brains. Also on GRPC move the logic to the network Node implementaiton.

File: nucliadb_node/nucliadb_node/shadow_shards.py (hunk 7 above)
Comment (lferran) [reply nucliadb_node/nucliadb_node/shadow_shards.py:180]: Due to this: https://stackoverflow.com/questions/54397600/new-line-character-in-serialized-messages

Some brain resources, when serialised to string, will include newline characters, which causes trouble for us when reading and writing to the file.

File: nucliadb_node/nucliadb_node/shadow_shards.py (hunk 8 above)
Comment (lferran) [reply nucliadb_node/nucliadb_node/shadow_shards.py:221]: mh I am not sure what you mean. Here we are returning all the operations that were written in a particular shadow shard. So we need to read the file.

File: nucliadb/nucliadb/ingest/service/writer.py (hunk 2 above)
Comment (lferran) [reply nucliadb/nucliadb/ingest/service/writer.py:747]: done

File: nucliadb/nucliadb/ingest/orm/node.py [hunk 9]
```diff
@@ -170,6 +171,98 @@ async def create_shard_by_kbid(cls, txn: Transaction, kbid: str) -> Shard:
...
+                updated_shards,
+                replica_id,
+                shadow_replica=ssresp.shard,
+                shadow_node=node_id,
+            )
+            if not updated:
+                raise UpdatingShardsError(f"Replica id not found: {replica_id}")
+            key = KB_SHARDS.format(kbid=kbid)
+            await txn.set(key, updated_shards.SerializeToString())
+            await txn.commit(resource=False)
```
Comment (bloodbare) [inline nucliadb/nucliadb/ingest/orm/node.py:211]: if txn is an argument you should not commit here.

File: nucliadb/nucliadb/ingest/orm/node.py [hunk 10]
```diff
@@ -170,6 +171,98 @@ async def create_shard_by_kbid(cls, txn: Transaction, kbid: str) -> Shard:
...
+            cleaned = cleanup_shadow_replica_from_shards(
+                updated_shards,
+                replica_id,
+            )
+            if not cleaned:
+                raise UpdatingShardsError()
+
+            key = KB_SHARDS.format(kbid=kbid)
+            await txn.set(key, updated_shards.SerializeToString())
+            await txn.commit(resource=False)
```
Comment (bloodbare) [inline nucliadb/nucliadb/ingest/orm/node.py:261]: Txn on argument

File: nucliadb/nucliadb/ingest/orm/node.py [hunk 11]
```diff
@@ -170,6 +171,98 @@ async def create_shard_by_kbid(cls, txn: Transaction, kbid: str) -> Shard:
...
+        except Exception:
+            await txn.abort()
```
Comment (bloodbare) [inline nucliadb/nucliadb/ingest/orm/node.py:263]: txn on argument

File: nucliadb/nucliadb/ingest/orm/node.py [hunk 12]
```diff
@@ -351,3 +444,62 @@ async def chitchat_update_node(members: List[ClusterMember]) -> None:
...
+
+
+def cleanup_shadow_replica_from_shards(shards: PBShards, replica_id: str) -> bool:
+    """
+    Returns whether it found the replica to cleanup in the shards proto message
+    """
+    replica = get_replica(shards, replica_id)
+    if replica is None:
+        return False
+    replica.has_shadow = False
```
Comment (bloodbare) [inline nucliadb/nucliadb/ingest/orm/node.py:471]: Has Shadow is duplicated of hasfield on replica?

File: nucliadb/nucliadb/ingest/service/writer.py [hunk 13]
```diff
@@ -741,6 +744,44 @@ async def generate_buffer(
...
+                txn, request.kbid, request.node, request.replica.id
+            )
+            response.success = True
+        except Exception as e:
+            event_id: Optional[str] = None
+            if SENTRY:
+                event_id = capture_exception(e)
+            logger.error(
+                f"Error creating shadow shard. Check sentry for more details. Event id: {event_id}"
+            )
```
Comment (bloodbare) [inline nucliadb/nucliadb/ingest/service/writer.py:764]: txn handling should be here

File: nucliadb/nucliadb/ingest/service/writer.py [hunk 14]
```diff
@@ -741,6 +744,44 @@ async def generate_buffer(
...
+            txn = await self.proc.driver.begin()
+            await node_klass.delete_shadow_shard(txn, request.kbid, request.replica.id)
+            response.success = True
+        except Exception as exc:
+            event_id: Optional[str] = None
+            if SENTRY:
+                event_id = capture_exception(exc)
+            logger.error(
+                f"Error deleting shadow shard. Check sentry for more details. Event id: {event_id}"
+            )
```
Comment (bloodbare) [inline nucliadb/nucliadb/ingest/service/writer.py:782]: txn handling should be here

**Review by bloodbare — CHANGES_REQUESTED** [review]

File: nucliadb/nucliadb/ingest/orm/node.py (hunk 12 above)
Comment (lferran) [reply nucliadb/nucliadb/ingest/orm/node.py:464]: 🤦🏼 totally. Still getting familiar with proto python api... fixing

File: nucliadb_node/nucliadb_node/shadow_shards.py [hunk 15]
```diff
@@ -0,0 +1,252 @@
...
+                async with aiofiles.open(path, mode="rb") as f:
+                    payload = await f.read()
+                    brain = Resource.FromString(payload)
+                    yield opcode, brain
+            elif opcode == OperationCode.DELETE:
+                yield opcode, uuid
+
+
+def get_shadow_shards_manager() -> ShadowShardsManager:
+    global SHADOW_SHARDS_MANAGER
```
Comment (bloodbare) [inline nucliadb_node/nucliadb_node/shadow_shards.py:241]: Could we switch to Dictionary instead of global ?

**Review by bloodbare — CHANGES_REQUESTED** [review]
Last one

**Review by bloodbare — APPROVED** [review]

**Pull request merged**